fv/netsets_test.go

```diff
@@ -0,0 +1,359 @@
+)
+
+const (
+	numPolicies    = 20
+	numNetworkSets = 100
+)
+
+var _ = FContext("Network sets connectivity tests with initialized Felix and etcd datastore", func() {
+
+})
+
+var _ = Context("Network sets churn test with initialized Felix and etcd datastore", func() {
+
+	var (
+		etcd     *containers.Container
+		felix    *containers.Container
+		felixPID int
+		client   client.Interface
+	)
+
+	getFelixPIDs := func() []int {
+		return felix.GetPIDs("calico-felix")
+	}
+
+	updateFelixPID := func() {
+		// Get Felix's PID.  This retry loop ensures that we don't get tripped up if we see multiple
+		// PIDs, which can happen transiently when Felix restarts/forks off a subprocess.
+		start := time.Now()
+		for {
+			pids := getFelixPIDs()
+			if len(pids) == 1 {
+				felixPID = pids[0]
+				break
+			}
+			Expect(time.Since(start)).To(BeNumerically("<", time.Second))
+			time.Sleep(50 * time.Millisecond)
+		}
+	}
+
+	BeforeEach(func() {
+		felix, etcd, client = containers.StartSingleNodeEtcdTopology(containers.TopologyOptions{
+			FelixLogSeverity: "info",
+		})
+		updateFelixPID()
+	})
+
+	AfterEach(func() {
+		if CurrentGinkgoTestDescription().Failed {
+			felix.Exec("iptables-save", "-c")
+		}
+		felix.Stop()
+
+		if CurrentGinkgoTestDescription().Failed {
+			etcd.Exec("etcdctl", "ls", "--recursive", "/")
+		}
+		etcd.Stop()
+	})
+
+	Describe("connectivity tests", func() {
+		var (
+			w  [4]*workload.Workload
+			cc *workload.ConnectivityChecker
+		)
+
+		createPolicy := func(policy *api.NetworkPolicy) {
+			log.WithField("policy", dumpPolicy(policy)).Info("Creating policy")
+			_, err := client.NetworkPolicies().Create(utils.Ctx, policy, utils.NoOptions)
+			Expect(err).NotTo(HaveOccurred())
+		}
+
+		BeforeEach(func() {
+			for ii := range w {
+				iiStr := strconv.Itoa(ii)
+				var ports string
+
+				ports = "3000"
+				w[ii] = workload.Run(
+					felix,
+					"w"+iiStr,
+					"cali0"+iiStr,
+					"10.65."+iiStr+".1", // IPs each in their own /24
+					ports,
+					"tcp",
+				)
+
+				w[ii].DefaultPort = "3000"
+				w[ii].Configure(client)
+			}
+
+			cc = &workload.ConnectivityChecker{
+				Protocol: "tcp",
+			}
+
+			pol := api.NewNetworkPolicy()
+			pol.Namespace = "fv"
+			pol.Name = "policy-1"
+			pol.Spec.Ingress = []api.Rule{
+				{
+					Action: "Allow",
+					Source: api.EntityRule{
+						Selector: "has(allow-as-source)",
+					},
+					Destination: api.EntityRule{
+						Selector: "has(allow-as-dest)",
+					},
+				},
+			}
+			pol.Spec.Selector = "all()"
+
+			createPolicy(pol)
+		})
+
+		It("with no matching network sets, should deny all", func() {
+			cc.ExpectNone(w[0], w[1])
+			cc.ExpectNone(w[0], w[2])
+			cc.ExpectNone(w[0], w[3])
+			cc.ExpectNone(w[1], w[0])
+			cc.ExpectNone(w[1], w[2])
+			cc.ExpectNone(w[1], w[3])
+			cc.ExpectNone(w[2], w[0])
+			cc.ExpectNone(w[2], w[1])
+			cc.ExpectNone(w[2], w[3])
+			cc.ExpectNone(w[3], w[0])
+			cc.ExpectNone(w[3], w[1])
+			cc.ExpectNone(w[3], w[2])
```

Comment:
Need to actually check the connectivity!
```
			Eventually(cc.ActualConnectivity, "10s", "100ms").Should(Equal(cc.ExpectedConnectivity()))
```

Reply:
(If I'm right, looks like this is missing throughout this file - so please check and fix throughout.)